### PR TITLE
enh(platformstatus): Save API Credentials and Proxy Options into remote database

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

While executing the script on a Remote
The Central API Credentials are now saved into the informations table and the proxy informations into the options table of the remote server. 
**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.04.x
- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x (master)

<h2> How this pull request can be tested ? </h2>
When executing the script on a remote server, the Central API Credentials should appears into the informations table and the proxy informations (if a proxy is used) should appear into the options table

## Checklist
- [x] I followed the **coding style guidelines** provided by Centreon
- [] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: bin/registerServerTopology-Func.php [hunk 1]
```diff
@@ -67,14 +66,28 @@ function registerRemote(string $ip, array $loginCredentials): array
     $topologyRepository = new \Centreon\Domain\Repository\TopologyRepository($db);
     $informationRepository = new \Centreon\Domain\Repository\InformationsRepository($db);
 
-    //hide menu
-    $topologyRepository->disableMenus();
+    $db->query(" SELECT * FROM `informations` WHERE `key` = 'isRemote' AND value = 'yes'");
+    $isRemote = $db->numberRows();
+    if (!$isRemote) {
+        require_once _CENTREON_PATH_ . "/src/Centreon/Infrastructure/CentreonLegacyDB/ServiceEntityRepository.php";
```
Comment (callapa) [inline bin/registerServerTopology-Func.php:72]: Why didn't you use the composer's autoloader directly?

**Review by loiclau — APPROVED** [review]

File: bin/registerServerTopology-Func.php [hunk 2]
```diff
@@ -97,6 +110,7 @@ function registerRemote(string $ip, array $loginCredentials): array
  */
 function hasRemoteChild(): bool
 {
+
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:113]: remove blank line

File: bin/registerServerTopology-Func.php [hunk 3]
```diff
@@ -125,18 +139,85 @@ function getChildren(CentreonDB $db): array
...
+                ];
+                break;
+            default:
+                $bindValues[':' . $key] = [
+                    \PDO::PARAM_STR => $value
+                ];
+                break;
+        }
+    }
+    $db->query("DELETE FROM informations WHERE `key` LIKE '%api%'");
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:175]:
```suggestion
    $db->query("DELETE FROM informations WHERE `key` LIKE 'api%'");
```

File: bin/registerServerTopology-Func.php [hunk 4]
```diff
@@ -125,18 +139,85 @@ function getChildren(CentreonDB $db): array
...
+                    break;
+                default:
+                    $bindProxyValues[':' . $key] = [
+                        \PDO::PARAM_STR => $value
+                    ];
+                    break;
+            }
+        }
+
+        $db->query("DELETE FROM options WHERE `key` LIKE '%proxy_%'");
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:210]:
```suggestion
        $db->query("DELETE FROM options WHERE `key` LIKE 'proxy_%'");
```

**Review by kduret — DISMISSED** [review]

**Review by loiclau — DISMISSED** [review]

File: bin/registerServerTopology-Func.php [hunk 5]
```diff
@@ -67,14 +66,28 @@ function registerRemote(string $ip, array $loginCredentials): array
     $topologyRepository = new \Centreon\Domain\Repository\TopologyRepository($db);
     $informationRepository = new \Centreon\Domain\Repository\InformationsRepository($db);
 
-    //hide menu
-    $topologyRepository->disableMenus();
+    $db->query(" SELECT * FROM `informations` WHERE `key` = 'isRemote' AND value = 'yes'");
```
Comment (callapa) [inline bin/registerServerTopology-Func.php:69]: Do a 'SELECT count(*)' instead of 'SELECT *'

File: bin/registerServerTopology-Func.php [hunk 6]
```diff
@@ -125,18 +138,85 @@ function getChildren(CentreonDB $db): array
...
+        $db->query("DELETE FROM options WHERE `key` LIKE 'proxy_%'");
+        $query = "INSERT INTO `options` (`key`, `value`) VALUES $queryValue";
+
+        $statement = $db->prepare($query);
+        foreach ($bindProxyValues as $token => $bindParams) {
+            foreach ($bindParams as $paramType => $paramValue) {
+                $statement->bindValue($token, $paramValue, $paramType);
+            }
+        }
+        $statement->execute();
```
Comment (callapa) [inline bin/registerServerTopology-Func.php:218]: Please use a transaction to cancel all changes in case of errors.

**Review by callapa — DISMISSED** [review]

**Review by loiclau — DISMISSED** [review]

File: bin/registerServerTopology-Func.php (hunk 6 above)
Comment (jeremyjaouen) [reply bin/registerServerTopology-Func.php:225]: done l.96

Comment (jeremyjaouen) [reply bin/registerServerTopology-Func.php:225]: done l.96

File: bin/registerServerTopology-Func.php [hunk 7]
```diff
@@ -125,18 +145,85 @@ function getChildren(CentreonDB $db): array
...
-    $statement->bindValue(':pwd', $loginCredentials['password'], \PDO::PARAM_STR);
+    foreach ($bindValues as $token => $bindParams) {
+        foreach ($bindParams as $paramType => $paramValue) {
+            $statement->bindValue($token, $paramValue, $paramType);
+        }
+    }
     $statement->execute();
+
+    if (!empty($proxyCredentials)) {
+        $queryValue = '';
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:192]:
```suggestion
        $queryValues = [];
```

File: bin/registerServerTopology-Func.php [hunk 8]
```diff
@@ -125,18 +145,85 @@ function getChildren(CentreonDB $db): array
...
+        $count = 1;
+        $bindProxyValues = [];
+        foreach ($proxyCredentials as $key => $value) {
+            if ($count === count($proxyCredentials)) {
+                $queryValue .= " ('$key', :$key)";
+            } else {
+                $queryValue .= " ('$key', :$key), ";
+            }
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:200]:
```suggestion
            $queryValues[] = " ('$key', :$key)";
```

File: bin/registerServerTopology-Func.php [hunk 9]
```diff
@@ -125,18 +145,85 @@ function getChildren(CentreonDB $db): array
...
+                default:
+                    $bindProxyValues[':' . $key] = [
+                        \PDO::PARAM_STR => $value
+                    ];
+                    break;
+            }
+        }
+
+        $db->query("DELETE FROM options WHERE `key` LIKE 'proxy_%'");
+        $query = "INSERT INTO `options` (`key`, `value`) VALUES $queryValue";
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:217]:
```suggestion
        $query = "INSERT INTO `options` (`key`, `value`) VALUES " . implode(',', $queryValues);
```

File: bin/registerServerTopology-Func.php [hunk 10]
```diff
@@ -125,18 +145,85 @@ function getChildren(CentreonDB $db): array
...
 
-
 /**
  * @param CentreonDB $db
  * @param array $loginCredentials
  */
 function registerCentralCredentials(CentreonDB $db, array $loginCredentials): void
 {
-    $query = "INSERT INTO `informations` (`key`, `value`) VALUES ('apiUsername', :username), ('apiCredentials', :pwd)";
+    $queryValue = '';
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:154]:
```suggestion
    $queryValues = [];
```

File: bin/registerServerTopology-Func.php [hunk 11]
```diff
@@ -125,18 +145,85 @@ function getChildren(CentreonDB $db): array
...
+    $proxyCredentials = $loginCredentials['proxy_informations'] ?? [];
+    $loginCredentials = array_filter($loginCredentials, function ($key) {
+        return $key !== 'proxy_informations';
+    }, ARRAY_FILTER_USE_KEY);
+    foreach ($loginCredentials as $key => $value) {
+        if ($count === count($loginCredentials)) {
+            $queryValue .= " ('$key', :$key)";
+        } else {
+            $queryValue .= " ('$key', :$key), ";
+        }
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:166]:
```suggestion
        $queryValues[] = " ('$key', :$key)";
```

File: bin/registerServerTopology-Func.php [hunk 12]
```diff
@@ -125,18 +145,85 @@ function getChildren(CentreonDB $db): array
...
+                break;
+            default:
+                $bindValues[':' . $key] = [
+                    \PDO::PARAM_STR => $value
+                ];
+                break;
+        }
+    }
+    $db->query("DELETE FROM informations WHERE `key` LIKE 'api%'");
+    $query = "INSERT INTO `informations` (`key`, `value`) VALUES $queryValue";
```
Comment (kduret) [inline bin/registerServerTopology-Func.php:182]:
```suggestion
    $query = "INSERT INTO `informations` (`key`, `value`) VALUES " . implode(',', $queryValues);
```

**Review by kduret — APPROVED** [review]

**Review by sc979 — APPROVED** [review]

**Pull request merged**